transports/janus_pfunix.c

```diff
@@ -0,0 +1,532 @@
+					JANUS_LOG(LOG_VERB, "Listening on Unix Socket %s... (Admin API)\n", pfname);
+					if(listen(admin_pfd, 5) != 0) {
+						JANUS_LOG(LOG_FATAL, "Listening on Unix Socket %s (Admin API) failed: %d, %s\n", pfname, errno, strerror(errno));
+						close(admin_pfd);
+						admin_pfd = 0;
+					}
+				}
+			}
+		}
+	}
+	janus_config_destroy(config);
+	config = NULL;
+	if(pfd == 0 && admin_pfd == 0) {
+		JANUS_LOG(LOG_FATAL, "No Unix Sockets server started, giving up...\n");
+		return -1;	/* No point in keeping the plugin loaded */
+	}
+
+	/* Create a couple of hashtables for all clients */
+	clients = g_hash_table_new(NULL, NULL);
+	clients_by_fd = g_hash_table_new(NULL, NULL);
+	janus_mutex_init(&clients_mutex);
+
+	GError *error = NULL;
+	/* Start the Unix Sockets service thread */
+	if(pfd > 0 || admin_pfd > 0) {
+		pfunix_thread = g_thread_try_new("pfunix thread", &janus_pfunix_thread, NULL, &error);
+		if(!pfunix_thread) {
+			g_atomic_int_set(&initialized, 0);
+			JANUS_LOG(LOG_ERR, "Got error %d (%s) trying to launch the Unix Sockets thread...\n", error->code, error->message ? error->message : "??");
+			return -1;
+		}
+	}
+
+	/* Done */
+	g_atomic_int_set(&initialized, 1);
+	JANUS_LOG(LOG_INFO, "%s initialized!\n", JANUS_PFUNIX_NAME);
+	return 0;
+}
+
+void janus_pfunix_destroy(void) {
+	if(!g_atomic_int_get(&initialized))
+		return;
+	g_atomic_int_set(&stopping, 1);
+
+	/* Stop the service thread */
+	char c;
+	write(write_fd[1], &c, 1);
+	if(pfunix_thread != NULL) {
+		g_thread_join(pfunix_thread);
+		pfunix_thread = NULL;
+	}
+
+	g_atomic_int_set(&initialized, 0);
+	g_atomic_int_set(&stopping, 0);
+	JANUS_LOG(LOG_INFO, "%s destroyed!\n", JANUS_PFUNIX_NAME);
+}
+
+int janus_pfunix_get_api_compatibility(void) {
+	/* Important! This is what your plugin MUST always return: don't lie here or bad things will happen */
+	return JANUS_TRANSPORT_API_VERSION;
+}
+
+int janus_pfunix_get_version(void) {
+	return JANUS_PFUNIX_VERSION;
+}
+
+const char *janus_pfunix_get_version_string(void) {
+	return JANUS_PFUNIX_VERSION_STRING;
+}
+
+const char *janus_pfunix_get_description(void) {
+	return JANUS_PFUNIX_DESCRIPTION;
+}
+
+const char *janus_pfunix_get_name(void) {
+	return JANUS_PFUNIX_NAME;
+}
+
+const char *janus_pfunix_get_author(void) {
+	return JANUS_PFUNIX_AUTHOR;
+}
+
+const char *janus_pfunix_get_package(void) {
+	return JANUS_PFUNIX_PACKAGE;
+}
+
+gboolean janus_pfunix_is_janus_api_enabled(void) {
+	return pfd > 0;
+}
+
+gboolean janus_pfunix_is_admin_api_enabled(void) {
+	return admin_pfd > 0;
+}
+
+int janus_pfunix_send_message(void *transport, void *request_id, gboolean admin, json_t *message) {
+	if(message == NULL)
+		return -1;
+	if(transport == NULL) {
+		g_free(message);
+		return -1;
+	}
+	/* Make sure this is related to a still valid Unix Sockets session */
+	janus_pfunix_client *client = (janus_pfunix_client *)transport;
+	janus_mutex_lock(&clients_mutex);
+	if(g_hash_table_lookup(clients, client) == NULL || client->fd <= 0) {
```

Comment:
Not with the way opaque transports are handled as of now. This is something I triedto change/fix in the new reference counters branch, where the transport reference is wrapped pretty much as we do with plugins (a core struct that contains both ends of the rope), so that both the core and plugin can ref/unref it accordingly and it never gets freed while it's used. Since we don't have that now, and I can't just envisage a delayed free (due to some misshap the core might try and send a message several seconds after the transport has gone) this is the closest thing I could come up with for a sanity check.